design/baremetal-operator/bmo-part-of-capm3.md

```diff
@@ -0,0 +1,283 @@
+cover first three use cases from the list provided above. Until now, there are
+three deployment use cases exist, i.e. production - `default`, local-developer -
+`ironic-keepalived-config`, ironic running outside of the cluster -
+`ironic-outside-config` as visualized in tree structure below.
+
+## Current structure of baremetal-operator deployment folder
+
+```diff
+tree deploy/
+
+deploy/
+├── bmo-capm3
+│   ├── ironic_bmo_configmap.env
+│   ├── kustomization.yaml
+│   └── README.md
+├── crds
+│   ├── kustomization.yaml
+│   └── metal3.io_baremetalhosts_crd.yaml
+├── default
+│   ├── ironic_bmo_configmap.env
+│   └── kustomization.yaml
+├── ironic_ci.env
+├── ironic-keepalived-config
+│   ├── image_patch.yaml
+│   ├── ironic_bmo_configmap.env
+│   └── kustomization.yaml
+├── ironic-outside-config
+│   ├── ironic_bmo_configmap.env
+│   └── kustomization.yaml
+├── namespace
+│   ├── kustomization.yaml
+│   └── namespace.yaml
+├── operator
+│   ├── ironic
+│   │   ├── kustomization.yaml
+│   │   └── operator_ironic.yaml
+│   ├── ironic_keepalived
+│   │   ├── kustomization.yaml
+│   │   └── operator_ironic_keepalived.yaml
+│   └── no_ironic
+│       ├── kustomization.yaml
+│       └── operator.yaml
+├── rbac
+│   ├── kustomization.yaml
+│   ├── role_binding.yaml
+│   ├── role.yaml
+│   └── service_account.yaml
+└── role.yaml -> rbac/role.yaml
+
+```
+
+Our main goal is to bring back deploy scripts as close to operator-sdk tools to
+be able to smoothly migrate to the new layout (to generate new API versions) as
+well as leaving only BMO deployments in deploy folder, while taking ironic
+deployments outside of the folder and moving them to the root of the repository
+to the newly created `baremetal-operator/ironic-deployment` folder. The following
+directory tree visualize new structure of deploy folder in more detail:
+
+## Proposed structure of baremetal-operator deployment folder
+
+```diff
+tree deploy/
+
+deploy/
+├── crds
+│   ├── kustomization.yaml
+│   └── metal3.io_baremetalhosts_crd.yaml
+├── default
+│   ├── ironic_bmo_configmap.env
+│   ├── kustomization.yaml
+│   └── kustomizeconfig.yaml
+├── ironic_ci.env
+├── namespace
+│   ├── kustomization.yaml
+│   └── namespace.yaml
+├── operator
+│   ├── bmo.yaml
+│   └── kustomization.yaml
+├── rbac
+│   ├── kustomization.yaml
+│   ├── role_binding.yaml
+│   ├── role.yaml
+│   └── service_account.yaml
+└── role.yaml -> rbac/role.yaml
+
+```
+
+As we can see, the `deploy` directory has one top level folder for deployment,
+namely `default` and it deploys only baremetal-operator through kustomization
+file and uses kustomization config file for teaching kustomize where to look at
+when substituting variables. In addition, `crds`, `namespace` and `rbac` folders
+have their own kustomization and yaml files. The following directory tree
+visualizes a new structure of ironic-deployment folder in more detail:
+
+## Proposed structure of ironic-deployment folder
+
+```diff
+tree ironic-deployment/
```

Comment:
This seems fine. I assume you rejected the idea of putting these files in new subdirectories of `deploy/`?

Reply:
It was not meant to be that at all, If I am not wrong, once we had a discussion putting all ironic related code in a separate repo as it was already under deploy folder. We wanted to do a bit of a re-organization in `deploy/` so then we decided to move ironic part into the separate folder outside of  `deploy/`.